File: code/packages/fsharp/write-ahead-log/tests/CodingAdventures.WriteAheadLog.Tests/WriteAheadLogTests.fs
namespace CodingAdventures.WriteAheadLog.Tests

open System
open System.IO
open CodingAdventures.WriteAheadLog
open Xunit

type WriteAheadLogTests() =
    let tempPath () =
        Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid():N}.wal")

    let deleteIfExists path =
        if File.Exists(path) then
            File.Delete(path)

    [<Fact>]
    member _.VersionExists() =
        Assert.Equal("0.1.0", WriteAheadLogPackage.Version)

    [<Fact>]
    member _.AppendRecordWritesLengthPrefixedPayloads() =
        let path = tempPath ()

        try
            use writer = new WalWriter(path)
            writer.AppendRecord([| 1uy; 2uy; 3uy |])
            writer.AppendRecord([| 4uy; 5uy |])
            (writer :> IDisposable).Dispose()

            Assert.Equal<byte array>(
                [| 3uy; 0uy; 0uy; 0uy; 1uy; 2uy; 3uy; 2uy; 0uy; 0uy; 0uy; 4uy; 5uy |],
                File.ReadAllBytes(path)
            )
        finally
            deleteIfExists path

    [<Fact>]
    member _.ReaderReturnsRecordsInOrderThenNoneAtEnd() =
        let path = tempPath ()

        try
            use writer = new WalWriter(path)
            writer.AppendRecord([| 10uy; 20uy |])
            writer.AppendRecord([| 30uy |])
            (writer :> IDisposable).Dispose()

            use reader = new WalReader(path)

            Assert.Equal<byte array>([| 10uy; 20uy |], reader.ReadNext().Value)
            Assert.Equal<byte array>([| 30uy |], reader.ReadNext().Value)
            Assert.True(reader.ReadNext().IsNone)
        finally
            deleteIfExists path

    [<Fact>]
    member _.EmptyRecordsRoundTrip() =
        let path = tempPath ()

        try
            use writer = new WalWriter(path)
            writer.AppendRecord([||])
            (writer :> IDisposable).Dispose()

            use reader = new WalReader(path)

            Assert.Equal<byte array>([||], reader.ReadNext().Value)
            Assert.True(reader.ReadNext().IsNone)
        finally
            deleteIfExists path

    [<Fact>]
    member _.WriterAppendsToExistingLog() =
        let path = tempPath ()

        try
            use writer = new WalWriter(path)
            writer.AppendRecord([| 1uy |])
            (writer :> IDisposable).Dispose()

            use writer2 = new WalWriter(path)
            writer2.AppendRecord([| 2uy; 3uy |])
            (writer2 :> IDisposable).Dispose()

            use reader = new WalReader(path)

            Assert.Equal<byte array>([| 1uy |], reader.ReadNext().Value)
            Assert.Equal<byte array>([| 2uy; 3uy |], reader.ReadNext().Value)
            Assert.True(reader.ReadNext().IsNone)
        finally
            deleteIfExists path

    [<Fact>]
    member _.PartialLengthPrefixIsTreatedAsEndOfFile() =
        let path = tempPath ()

        try
            File.WriteAllBytes(path, [| 2uy; 0uy |])

            use reader = new WalReader(path)

            Assert.True(reader.ReadNext().IsNone)
        finally
            deleteIfExists path

    [<Fact>]
    member _.TruncatedPayloadThrows() =
        let path = tempPath ()

        try
            File.WriteAllBytes(path, [| 4uy; 0uy; 0uy; 0uy; 1uy; 2uy |])

            use reader = new WalReader(path)

            Assert.Throws<EndOfStreamException>(fun () -> reader.ReadNext() |> ignore) |> ignore
        finally
            deleteIfExists path

    [<Fact>]
    member _.InvalidArgumentsAreRejected() =
        Assert.Throws<ArgumentException>(fun () -> new WalWriter("") |> ignore) |> ignore
        Assert.Throws<ArgumentException>(fun () -> new WalReader("") |> ignore) |> ignore
        Assert.Throws<FileNotFoundException>(fun () -> new WalReader(Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"))) |> ignore) |> ignore

    [<Fact>]
    member _.WriterRejectsNullPayload() =
        let path = tempPath ()

        try
            use writer = new WalWriter(path)

            Assert.Throws<ArgumentNullException>(fun () -> writer.AppendRecord(Unchecked.defaultof<byte array>)) |> ignore
        finally
            deleteIfExists path

File: code/packages/csharp/write-ahead-log/tests/CodingAdventures.WriteAheadLog.Tests/WriteAheadLogTests.cs
using CodingAdventures.WriteAheadLog;

namespace CodingAdventures.WriteAheadLog.Tests;

public sealed class WriteAheadLogTests
{
    [Fact]
    public void VersionExists()
    {
        Assert.Equal("0.1.0", WriteAheadLogPackage.Version);
    }

    [Fact]
    public void AppendRecordWritesLengthPrefixedPayloads()
    {
        var path = TempPath();
        try
        {
            using (var writer = new WalWriter(path))
            {
                writer.AppendRecord([1, 2, 3]);
                writer.AppendRecord([4, 5]);
            }

            Assert.Equal([3, 0, 0, 0, 1, 2, 3, 2, 0, 0, 0, 4, 5], File.ReadAllBytes(path));
        }
        finally
        {
            DeleteIfExists(path);
        }
    }

    [Fact]
    public void ReaderReturnsRecordsInOrderThenNullAtEnd()
    {
        var path = TempPath();
        try
        {
            using (var writer = new WalWriter(path))
            {
                writer.AppendRecord([10, 20]);
                writer.AppendRecord([30]);
            }

            using var reader = new WalReader(path);

            Assert.Equal([10, 20], reader.ReadNext());
            Assert.Equal([30], reader.ReadNext());
            Assert.Null(reader.ReadNext());
        }
        finally
        {
            DeleteIfExists(path);
        }
    }

    [Fact]
    public void EmptyRecordsRoundTrip()
    {
        var path = TempPath();
        try
        {
            using (var writer = new WalWriter(path))
            {
                writer.AppendRecord([]);
            }

            using var reader = new WalReader(path);

            var record = reader.ReadNext();
            Assert.NotNull(record);
            Assert.Empty(record);
            Assert.Null(reader.ReadNext());
        }
        finally
        {
            DeleteIfExists(path);
        }
    }

    [Fact]
    public void WriterAppendsToExistingLog()
    {
        var path = TempPath();
        try
        {
            using (var writer = new WalWriter(path))
            {
                writer.AppendRecord([1]);
            }

            using (var writer = new WalWriter(path))
            {
                writer.AppendRecord([2, 3]);
            }

            using var reader = new WalReader(path);

            Assert.Equal([1], reader.ReadNext());
            Assert.Equal([2, 3], reader.ReadNext());
            Assert.Null(reader.ReadNext());
        }
        finally
        {
            DeleteIfExists(path);
        }
    }

    [Fact]
    public void PartialLengthPrefixIsTreatedAsEndOfFile()
    {
        var path = TempPath();
        try
        {
            File.WriteAllBytes(path, [2, 0]);

            using var reader = new WalReader(path);

            Assert.Null(reader.ReadNext());
        }
        finally
        {
            DeleteIfExists(path);
        }
    }

    [Fact]
    public void TruncatedPayloadThrows()
    {
        var path = TempPath();
        try
        {
            File.WriteAllBytes(path, [4, 0, 0, 0, 1, 2]);

            using var reader = new WalReader(path);

            Assert.Throws<EndOfStreamException>(() => reader.ReadNext());
        }
        finally
        {
            DeleteIfExists(path);
        }
    }

    [Fact]
    public void InvalidArgumentsAreRejected()
    {
        Assert.Throws<ArgumentException>(() => new WalWriter(""));
        Assert.Throws<ArgumentException>(() => new WalReader(""));
        Assert.Throws<FileNotFoundException>(() => new WalReader(Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"))));
    }

    private static string TempPath()
    {
        return Path.Combine(Path.GetTempPath(), $"{Guid.NewGuid():N}.wal");
    }

    private static void DeleteIfExists(string path)
    {
        if (File.Exists(path))
        {
            File.Delete(path);
        }
    }
}
